### ISSUE TITLE
[FEATURE] Replace static dev ribbon image with a dynamic CSS based non-prod ribbon

### ISSUE THREAD

**Issue by ddewaele** [jhipster/generator-jhipster#3368]
Context : 

- Currently jHipster has a static development ribbon image that is displayed when initially running with the dev profile
- When building with the -Pprod profile, at build time Gulp ensures that the development ribbon is removed.
- (doubt) I think that at that point the ```app.constants.js``` is modified and can never be generated to get the development ribbon back (you need the -Pdev profile to change it back, but -Pdev will not run gulp).
- The docs always talk about 2 profiles, but in some environments where there are multiple non-prod environments, having jHipster (optionally) show a banner to clearly visualise what non-prod environment we're on (and if we have the correct spring profile on a particular environment) is can be useful.

Proposal of this feature : 

- Remove the static development ribbon image with a [CSS based ribbon](https://github.com/simonwhitaker/github-fork-ribbon-css) so that is more dynamic.
- When having multiple non-prod environments, (dev, tst, uat, ....), we could show appropriate banner to indicate with what spring profile the app has been started.  (can be optional)
- For the prod profile we can remove the banner

Thinking about implementing this for our projects, if you feel this could be useful for jHipster I'd be happy to create a PR

**Comment by gmarziou** [comment]
Good idea.

Another benefit of using CSS is that we could translate it in all languages we support.

Could you clarify how you would switch between profiles? I imagine doing it server-side maybe using Thymeleaf for injecting profile as a template variable.

**Comment by ddewaele** [comment]
Possibly, need to look at the specifics. It's only on the index page and that's only loaded once normally ...  will draft a proposal here this weekend. But need some r&r now :)

**Comment by deepu105** [comment]
I also did this in one of my project, I had a simple rest controller which
would return active profiles and this endpoint was not protected so that it
is available all the time, and had a directive and service which would use
this to set the ribbon text. But i didn't consider it a very clean solution
and that's why didn't add it to the generator itself.

**Comment by gmarziou** [comment]
Yes this may be the ricky parts and also what to display when there are multiple profiles active at same time.

Anyway, even with only the CSS paft, we would benefit from i18n.

**Comment by ddewaele** [comment]
Been playing around with this a bit with one of my projects.

**jHipster ribbon config option (application.yml)**

Instructs jHipster to show the ribbon for the provided profiles :

```
ribbon:
   displayOnActiveProfiles: dev
```

If users have additional envs all they need to do is this : 

```
ribbon:
   displayOnActiveProfiles: dev,tst,uat
```

**i18n**

labels can be provided in global.json

```json
    "ribbon": {
    	"dev":"Development",
    	"tst":"Test",
    	"uat":"User Acceptance"
    }
```

A controller checks the active spring boot profiles :

- If it matches with one of the ribbon profiles it can inject a ```ribbonEnv``` template var (```ribbon.dev | ribbon.tst | ...```). 
- If no match is found we don't inject and don't show the ribbon.
- Unlikely, but if multiple matches are found we return the first match.


With Tymeleaf it would look something like this : 

```
<div th:if="${ribbonEnv != null}" th:classappend="${ribbonEnv}" class="github-fork-ribbon left-top" translate="" data-th-attr="translate-attr-title='ribbon.' + ${ribbonEnv}"></div>
```

**Comment by cicoub13** [comment]
:+1

**Comment by ddewaele** [comment]
If you think this would be useful to put in the generator I can create a PR so it can be reviewed.

Some remarks:

- Injecting the template var into the page is currently done server-side, meaning that the index.html needs to be served by a Spring controller. Is that ok for you guys or do you prefer to do it on the UI (unsecured API call to return the profile/environment info)
- Current gulp ENV variable processing with ```app.constants.js``` (currently used to show/hide the ribbon) will remain as-is as it's also used for other stuff (show/hide API / Database links, enable/disable angular debug data,...)

**Comment by deepu105** [comment]
@ddewaele I like everything except the thyme-leaf injection. I would prefer the API call. I dont know what others will prefer.

I actually dont like the `ENV` in `app.constants.js` as you said in actual projects there could be more profiles and we now have spring profile switches as well.

I would recommend having an unsecured endpoint which returns all the active profiles and on the client side a simple service to get this. Does anybody think that there is a security concern here? I might not have thought through completely

This way we can achieve the purpose of this ticket as well as more

1. dynamically show ribbon base on profile
2. show hide DB and swagger page based on profile and switches
3. remove the environment from `app.constants.js` which cannot be always guaranteed to be correct.

**Comment by ddewaele** [comment]
I'll leave it up to you guys to decide what is better. Both work for me.

For removing the ENV from app.constants.js, I initially put that out of scope for this issue as we need to have a clear understanding of the build profiles VS spring profiles and how they are used in jHipster.  That's a whole discussion in itself. 

Currently jHipster supports 

**2 build profiles (dev / prod).**

- This is very static. 
- Once your app is built it can no longer be changed. 
- At build time used to minify the JS
- At run time used (via the ENV constant) to enable/disable certain UI stuff (admin links / angular debug data / ...). Cannot be changed at runtime for a given WAR.
- "prod" is a magic word used to switch on/off certain things. 

**2 spring profiles (dev / prod).**

- This is more dynamic. 
- The same WAR (build with a prod build profile) can be used with different spring profiles.

If don't really like these 2 prod concepts as they can be conflicting and confusing, but we would still need _"something"_ at runtime to determine if we want to

- show/hide database / swagger
- enable angular debug data

Would that "something" become the "prod" (magic word) spring profile instead of the current static ENV from app.constants.js ?

**Comment by deepu105** [comment]
Yes today we have 2 build profiles and 2 spring profiles and in a way they are tied tightly together.

if you build using `dev` you can only run using the same spring profile, if you try to run it with `prod` it will not work. Same goes for `prod` profile so in a way we can see it as just 2 profiles and need not really separate them out in theory.

We currently have some dynamic spring profiles `no-swagger` and `no-liquibase` which can be used along with either `prod` or `dev` to disable swagger and liquibase respectively.

currently to the users the `dev` and `prod` profiles are straight forward to use IMO and with https://github.com/jhipster/generator-jhipster/pull/3224 it will be even more easy to use.

so ya based on profiles we would need to do below

1. show development ribbon if profile doesnt have `prod`
2. hide swagger link in navbar if profile has `no-swagger`
3. hide database link in navbar if profile has `prod`
4. disable angular debug if profile has `prod`

**Comment by ddewaele** [comment]
ok perfect works for me. 

To summarize :  

- ENV will be removed from app.constants
- Maven/gradle dev/prod build profile will do nothing more than enable/disable gulp during packaging.
- All the rest will be done through spring profiles.
- We make an unsecured api endpoint that returns the active spring profiles

Questions :

- How does the client retrieve the ribbon config (ribbon.displayOnActiveProfiles - to know if the ribbon needs to be shown for the given active profiles) ? This is currently in the yml file. Angular cannot access it (unsecured) I guess.
- Is everybody ok with the unauthenticated endpoint ?

**Comment by deepu105** [comment]
for your first question you can access it in the spring controller using environment and sent it as a param in the api response object.

for your second question Im not sure @jhipster/developers does anyone have an issue with whats proposed above? or does anyone have a better approach?

**Comment by gmarziou** [comment]
OK for me, if someone wants to secure it, it is easy to edit SecurityConfiguration.

Also, I'd like to specify that if the server fails to return the active profiles for some reason, the default behavior should be to hide the ribbon.

**Comment by deepu105** [comment]
@ddewaele since no body has any concern you can proceed with the api endpoint approach

**Comment by ddewaele** [comment]
@deepu105 : have reserved time for this this weekend so you can expect a PR soon.

**Comment by ddewaele** [comment]
Before I create a PR, here are some comments on how this could be implemented. Feedback welcome. If you want me to create a PR so we can discuss it there let me know....

### Services

#### Spring Resource

An ActiveProfiles resource is exposed on ```/api/activeProfiles``` that returns a json structure containing:

- 1 or more activeProfiles
- the ribbonEnv (if there's a match between activeProfiles and ribbonEnvs).


```json
{
	"activeProfiles":["dev"],
	"ribbonEnv":"dev"
}
```

The Spring Resource:


```java
package be.ixor.metro.assetmgmt.web.rest;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import javax.inject.Inject;

import org.springframework.core.env.Environment;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import be.ixor.metro.assetmgmt.config.JHipsterProperties;

@RestController
@RequestMapping("/api")
public class ActiveProfilesResource {

	@Inject
	Environment env;
	
    @Inject
    private JHipsterProperties jHipsterProperties;

	@RequestMapping("/activeProfiles")
	public ActiveProfileResponse getActiveProfiles() {
		return new ActiveProfileResponse(env.getActiveProfiles(), getRibbonEnv());
	}
	
	private String getRibbonEnv() {
		String[] activeProfiles = env.getActiveProfiles();
		List<String> ribbonProfiles = new ArrayList<>(Arrays.asList(jHipsterProperties.getRibbon().getDisplayOnActiveProfiles()));
		List<String> springBootProfiles = Arrays.asList(activeProfiles);
    	ribbonProfiles.retainAll(springBootProfiles);
    	
    	if (ribbonProfiles.size()>0) {
    		return ribbonProfiles.get(0);
    	}
		return null;
	}
	
	class ActiveProfileResponse {
		
		public String[] activeProfiles;
		public String ribbonEnv;
		
		ActiveProfileResponse(String[] activeProfiles,String ribbonEnv) {
			this.activeProfiles=activeProfiles;
			this.ribbonEnv=ribbonEnv;
		}
		
	}
	
}
```


### Angular

An Angular service is created to call the service above.

```javascript
(function() {
    'use strict';

    angular
        .module('assetmgmtApp')
        .service('ActiveProfiles', ActiveProfiles);

    ActiveProfiles.$inject = ['$resource'];

    function ActiveProfiles ($resource) {
        var service = $resource('api/activeProfiles', {}, {
            'get': { method: 'GET', params: {}, isArray: false,
                interceptor: {
                    response: function(response) {
                        // expose response
                        return response;
                    }
                }
            }
        });

        return service;
    }
})();
```


## pageRibbon directive

An Angular directive is placed on the page.

```html
    <link rel="stylesheet" href="bower_components/github-fork-ribbon-css/gh-fork-ribbon.css" />

	<page-ribbon></page-ribbon>
```

- It is responsible for showing the ribbon, depending on the activeProfile.
- It hides the ribbon when no ribbonEnv is matched (ex: prod).
- It adds the ribbonEnv as a class for CSS styling (different colors for difference environments)

```javascript
(function() {
    'use strict';

    angular
        .module('assetmgmtApp')
        .directive('pageRibbon', pageRibbon);

    function pageRibbon(ActiveProfiles,$translate,$rootScope) {
        var directive = {
    		replace: true,
            restrict: 'AE',
            template: '<div class="github-fork-ribbon left-top" title="{{ribbonTitle}}"></div>',
            link: linkFunc
        };

        return directive;

        function linkFunc(scope, element, attrs) {
            ActiveProfiles.get(function(result) {
                if (result.data.ribbonEnv) {
                	$rootScope.$on('$translateChangeSuccess', function () {
                	    scope.ribbonTitle = $translate.instant('ribbon.' + result.data.ribbonEnv);
                	});
                	element.addClass(result.data.ribbonEnv);
                } else {
                	element.addClass('hidden');
                }
            });
        }
    }
})();
```

## Security Configuration

We make the API call unprotected in ```SecurityConfiguration```

```
.antMatchers("/api/activeProfiles").permitAll()
```

## i18n support

In global.json we provide 3 default values. By default we don't show a ribbon for prod.

```json
    "ribbon": {
    	"dev":"Development",
    	"test":"Test",
    	"uat":"User Acceptance"
    }
```



Questions :

- Currently an ENV constant is passed around with a static value. Can this be replaced with live calls to the ```api/activeProfiles``` or should this also be cached somewhere on the UI ?
- Do we want a dependency to [github-fork-ribbon-css](https://github.com/simonwhitaker/github-fork-ribbon-css) or do we want to create a ribbon ourselves ?
- There's still [a pending PR](https://github.com/simonwhitaker/github-fork-ribbon-css/pull/48) to fix an alignment issue with Twitter Bootstrap

**Comment by deepu105** [comment]
Looks ok to me

To answer your questions
1. We can replace the environment with the data from this call, but needs to be cached as you say, may be as an angular service variable. In the service you can assign the result to a var and check for it before making the server call, similar to Principal service
2. I guess for few lines of CSS we can just have it in our css withe a reference link to the repo. I would remove some of the gradient css and stuff and make it a minimal flat ribbon to just serve the purpose.

**Comment by gmarziou** [comment]
> a reference link to the repo

Which repo?

**Comment by ddewaele** [comment]
@gmarziou  : [github-fork-ribbon-css](https://github.com/simonwhitaker/github-fork-ribbon-css)

**Comment by gmarziou** [comment]
OK. Well it's MIT license and we don't need the link part, so we could fork it and simplify it though giving credits to original author.

**Comment by jdubois** [comment]
No don't copy code from another repo: we either depend on it using Bower (favorite solution, always) or code our own.
Don't forget that once the code is generated it will have the user's licence.

**Comment by deepu105** [comment]
I dont want to depend on it coz its just few lines of css, so what I meant
was we copy some of the css required to our main.css file and just add a
comment crediting the original author with a link like we already do for
some of the css we copy from other places.

**Comment by ddewaele** [comment]
I'm also somewhat reluctant to have a bower dependency on such a small package. It still has an open PRs that need to be merged before we can use it and it's very Github specific now (intended as a fork-me github ribbon so it has github in the class names)

My suggestion would also be to : 

- Include it in the jHipster code (part of main.css or a specific ribbon.css).
- Credit the author (inspired by....)
- Do the bugfixing , change the style class names (remove the ```github-``` reference) and optimise where we can.

@deepu105 : Where else in jHipster have you pulled in CSS from other authors ? Does that end up in main.css or would you like a seperate css for the ribbon ?

**Comment by jdubois** [comment]
It's a license issue: when one of our users has generated an app, he expects the generated code to be his. From a legal point of view, it's a bit more compex, but it could be argued we do the same thing as an IDE: when you do a "generate getters and setters" with Eclipse that code is also yours.
We've already had that issue in the past, with code copied from ehcache, and that's not a good memory because I had to change everything.
So I don't want any code copy/pasted on the project. It's a legal issue for our users.

**Comment by deepu105** [comment]
@ddewaele in that case since its generic CSS and not any IPR, we can write our own css classes to do the same. you can use the github-fork-ribbon-css as reference if needed. Anyway we don't need the ribbon to look that specific to a github fork me button. we just need a flat basic ribbon. lets settle it that way with our own css in main.css file.

**Comment by ddewaele** [comment]
No worries ... Found a [nice guide](https://unindented.org/articles/github-ribbon-using-css-transforms/). Will prep the PR today so you can review. Still need to do some testing.

**Comment by deepu105** [comment]
closed via #3474 
@ddewaele plz add the fix #issuenumber in commit for autoclose next timeplease
